### PR TITLE
Geyser extension loading

### PR DESCRIPTION
**Pull request by ImDaBigBoss**
I thought that I might as well get a start on Geyser extensions, so here they are. I've currently not actually implemented anything, but you can access `GeyserImpl` from the plugin though `this.getGeyser()`.
Please notify me if there are any functions you want to rename or move. And for the project maintainers, feel free to push any changes that you want.

You can find an example extension here: [https://github.com/ImDaBigBoss/TestGeyserExtension](https://github.com/ImDaBigBoss/TestGeyserExtension)

Things that probably need doing:
-  Event listeners

I'm open to more suggestions.

### CODE REVIEW

**Review by Camotoy — DISMISSED** [review]
Immediate thought: all this should be moved to the `api/geyser` module. `core` should contain zero API (think of it like NMS).

**Comment by ImDaBigBoss** [comment]
So I shouldn't link anything in there to the `core`module. I'll do that.

**Comment by ImDaBigBoss** [comment]
I had started a list of things that should probably need doing, but that’s something for another pull request, for the API in general.

File: api/geyser/src/main/java/org/geysermc/geyser/api/extension/Extension.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * @link https://github.com/GeyserMC/Geyser
+ */
+
+package org.geysermc.geyser.api.extension;
+
+import org.geysermc.api.GeyserApiBase;
+import java.io.File;
+import java.io.InputStream;
+
+public interface Extension {
```
Comment (Camotoy) [inline api/geyser/src/main/java/org/geysermc/geyser/api/extension/Extension.java:32]: This all needs to be Javadoc'd.

File: api/geyser/src/main/java/org/geysermc/geyser/api/extension/ExtensionDescription.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ * @author GeyserMC
+ * @link https://github.com/GeyserMC/Geyser
+ */
+
+package org.geysermc.geyser.api.extension;
+
+import java.util.*;
+
+public interface ExtensionDescription {
```
Comment (Camotoy) [inline api/geyser/src/main/java/org/geysermc/geyser/api/extension/ExtensionDescription.java:30]: Javadocs

**Review by Camotoy — COMMENTED** [review]
The only implementation that I think is essential before merge is that all extensions are printed in a Geyser dump.

File: api/geyser/src/main/java/org/geysermc/geyser/api/extension/ExtensionDescription.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    String name();
+    String main();
+    List<String> ApiVersions();
```
Comment (RednedEpic) [inline api/geyser/src/main/java/org/geysermc/geyser/api/extension/ExtensionDescription.java:33]:
```suggestion
    String apiVersion();
```
Should just be a single API version which signifies the lowest supported API version.

**Review by RednedEpic — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/geysermc/geyser/extension/GeyserExtensionDescription.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.*;
+
+public class GeyserExtensionDescription implements org.geysermc.geyser.api.extension.ExtensionDescription {
+    private String name;
+    private String main;
+    private String api;
+    private String version;
+    private final List<String> authors = new ArrayList<>();
+
+    public GeyserExtensionDescription(String yamlString) throws InvalidDescriptionException {
```
Comment (Konicai) [inline core/src/main/java/org/geysermc/geyser/extension/GeyserExtensionDescription.java:40]: Yaml can read from an inputstream afaik

File: core/src/main/java/org/geysermc/geyser/dump/DumpInfo.java [hunk 5]
```diff
@@ -277,4 +282,15 @@ public LogsInfo() {
...
+
+    @Getter
+    @AllArgsConstructor
+    public static class ExtensionInfo {
+        public boolean enabled;
+        public String name;
+        public String version;
+        public String main;
+        public List<String> authors;
+        public String apiVersion;
```
Comment (Konicai) [inline core/src/main/java/org/geysermc/geyser/dump/DumpInfo.java:294]: I would put api version below version

File: core/src/main/java/org/geysermc/geyser/extension/GeyserExtensionLoader.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                }
+            }
+            return clazz;
+        } catch(NullPointerException s) {
+            return null;
+        }
+    }
+
+    void setClass(String name, final Class<?> clazz) {
+        if(!classes.containsKey(name)) {
```
Comment (Konicai) [inline core/src/main/java/org/geysermc/geyser/extension/GeyserExtensionLoader.java:154]: Spacing

File: api/geyser/src/main/java/org/geysermc/geyser/api/extension/GeyserExtension.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+     * @return true if the extension is enabled
+     */
+    public boolean isEnabled() {
+        return this.enabled;
+    }
+
+    /**
+     * Gets if the extension is enabled
+     *
+     * @return true if the extension is enabled
```
Comment (Konicai) [inline api/geyser/src/main/java/org/geysermc/geyser/api/extension/GeyserExtension.java:80]: inaccurate javadoc (no return and modifies state)

File: core/src/main/java/org/geysermc/geyser/GeyserImpl.java [hunk 8]
```diff
@@ -154,6 +155,8 @@ private GeyserImpl(PlatformType platformType, GeyserBootstrap bootstrap) {
         MessageTranslator.init();
         MinecraftLocale.init();
 
+        GeyserExtensionManager.init();
```
Comment (Konicai) [inline core/src/main/java/org/geysermc/geyser/GeyserImpl.java:158]: If extensions are not enabled in start() then a reload caused by the server will permanently disable extensions, afaik. 

I would remove the enabling of extensions from extensionmanager#init() and then enable them all in GeyserImpl#start()

File: core/src/main/java/org/geysermc/geyser/extension/GeyserExtensionClassLoader.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        return this.findClass(name, true);
+    }
+
+    protected Class<?> findClass(String name, boolean checkGlobal) throws ClassNotFoundException {
+        if (name.startsWith("org.geysermc.geyser.") || name.startsWith("org.geysermc.connector.") || name.startsWith("net.minecraft.")) {
+            throw new ClassNotFoundException(name);
+        }
+        Class<?> result = classes.get(name);
+        if(result == null) {
+            if(checkGlobal) {
```
Comment (Konicai) [inline core/src/main/java/org/geysermc/geyser/extension/GeyserExtensionClassLoader.java:82]: Spacing, here and just below

File: api/geyser/src/main/java/org/geysermc/geyser/api/extension/GeyserExtension.java (hunk 7 above)
Comment (ImDaBigBoss) [reply api/geyser/src/main/java/org/geysermc/geyser/api/extension/GeyserExtension.java:80]: That was me moving javadocs out of the old Extension class. I must have copied the wrong one without realising.

**Comment by ImDaBigBoss** [comment]
Something else that needs doing is I need to setup log message translations. Could I please get some information on how to do so?

**Review by RednedEpic — APPROVED** [review]

**Comment by iZeaoGamer** [comment]
I’m gay.

**Pull request merged**